lampe/inference.py

```diff
@@ -524,14 +540,183 @@ def forward(self, theta: Tensor, x: Tensor) -> Tensor:
+        moments: The parameters moments :math:`\mu` and :math:`\sigma`. If provided,
+            the moments are used to standardize the parameters.
+        build: The network constructor (e.g. :class:`lampe.nn.ResMLP`).
+        kwargs: Keyword arguments passed to the constructor.
+    """
+
+    def __init__(
+        self,
+        theta_dim: int,
+        x_dim: int,
+        moments: Tuple[Tensor, Tensor] = None,
+        build: Callable[[int, int], nn.Module] = MLP,
+        **kwargs,
+    ):
+        super().__init__()
+
+        if moments is None:
+            self.standardize = nn.Identity()
+        else:
+            mu, sigma = moments
+            self.standardize = Affine(-mu / sigma, 1 / sigma, trainable=False)
+
+        self.net = build(theta_dim + x_dim + 2, theta_dim, **kwargs)
+
+        self.register_buffer('zeros', torch.zeros(theta_dim))
+        self.register_buffer('ones', torch.ones(theta_dim))
+
+    def forward(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        r"""
+        Arguments:
+            theta: The parameters :math:`\theta`, with shape :math:`(*, D)`.
+            x: The observation :math:`x`, with shape :math:`(*, L)`.
+            t: The time :math:`t`, with shape :math:`(*,).`
+
+        Returns:
+            The score :math:`s_\phi(\theta, x, t)`, with shape :math:`(*,)`.
+        """
+
+        theta = self.standardize(theta)
+        t = torch.stack((
+            torch.cos(math.pi * t),
+            torch.sin(math.pi * t),
+        ), dim=-1)
+        theta, x, t = broadcast(theta, x, t, ignore=1)
+
+        return self.net(torch.cat((theta, x, t), dim=-1))
+
+    def alpha(self, t: Tensor) -> Tensor:
+        return torch.exp(-8.0 * t**2)
+
+    def beta(self, t: Tensor) -> Tensor:
+        return 16.0 * t
+
+    def ode(self, theta: Tensor, x: Tensor, t: Tensor) -> Tensor:
+        drift = -self.beta(t) / 2 * theta
+        diffusion = -self.beta(t) / 2 * (1 + self.alpha(t)) * self.forward(theta, x, t)
+
+        return drift + diffusion
+
+    def flow(self, x: Tensor) -> Distribution:
+        r"""
+        Arguments:
+            x: The observation :math:`x`, with shape :math:`(*, L)`.
+
+        Returns:
+            The posterior distribution :math:`p_\phi(\theta | x)` induced by the
+            probability flow ODE.
+
+        Note:
+            The :func:`log_prob` method of the returned distribution is an unbiased
+            stochastic estimator of the true log-density. One should average over
+            a sufficient number of evaluations to attain small errors.
+        """
+
+        batch_shape = x.shape[:-1]
+
+        return NormalizingFlow(
+            transform=FFJTransform(
+                f=lambda theta, t: self.ode(theta, x, t),
+                time=x.new_tensor(1.0),
+                phi=(x, *self.parameters()),
+            ),
+            base=DiagNormal(self.zeros, self.ones).expand(batch_shape),
+        )
```

Comment:
I am not sure if there is any definite API yet but I find a bit inconsistent that `NPE` defines a `sample` method based on its own `self.flow` and here there is no `sample` method implemented.

Reply:
Yes, I dislike it as well, but I am not sure of the best way to sample from the trained estimator yet. From my toy experiments sampling from the probability flow ODE gives good samples and is pretty fast, but Song et al. (2021) put forward some evidence that ancestral and reverse SDE sampling yields more qualitative samples.

Reply:
Yeah, I was more worried about the lack of consistency between the interfaces of the classes than about the particular implementation of the `sample` interface.